lib/elixir/src/elixir_expand.erl

```diff
@@ -456,6 +456,11 @@ expand(Pid, S, E) when is_pid(Pid) ->
       {Pid, E}
   end;
 
+expand(Zero, S, #{context := match} = E) when is_float(Zero), Zero == 0.0 ->
+  elixir_errors:file_warn([], E, ?MODULE, invalid_match_on_zero_float),
+  Escaped = {{'.', [], [erlang, '+']}, [], [0.0]},
```

Comment:
Shouldn't this consider if we have -0.0 or 0.0 accordingly?

Reply:
Also, this current implementation makes it so `+0.0` repeats the plus operator twice in the Erlang AST, right?

Reply:
In order to address these, I think you will pretty much need to add the same code I did on elixir_erl and expand_remote_arg. Which basically means our code is ultimately the same but moved to different places.

With that said, I would prefer to keep the AST smaller (a single float) than bigger. So, if you agree with the assessment above, let's merge my PR and then you push your pattern tests fixes?

Reply:
> Shouldn't this consider if we have -0.0 or 0.0 accordingly?

Yes, definitely 👍

> Also, this current implementation makes it so +0.0 repeats the plus operator twice in the Erlang AST, right?

As far as I checked, it doesn't, this piece only gets called for `0.0`, not `+0.0`. But https://github.com/elixir-lang/elixir/pull/12950 does repeat it: `+0.0` becomes `{:op, 4, :+, {:op, 4, :+, {:float, 4, 0.0}}}` outside of matches, and `-0.0` becomes `-(+0.0)`.

> With that said, I would prefer to keep the AST smaller (a single float) than bigger.

I understand, but in this case, shouldn't we do it for non matches too? Right now, only matches are smaller (which is a bit ironic given that it is the only place we would need the sign info 😉).

> So, if you agree with the assessment above, let's merge my PR and then you push your pattern tests fixes?

Reply:
> I understand, but in this case, shouldn't we do it for non matches too? 

I am fine with always doing it. :) Can you please amend this PR then so we always convert to -0.0 or +0.0 as literals instead of AST and we build the AST back in the Erlang pass? I will close mine. Awesome job!

Reply:
@josevalim thank you!

I'm not 100% sure I understood, is [6ac0664](https://github.com/elixir-lang/elixir/pull/12949/commits/6ac066462817f6e5a9041d25a58cb734ccf148dd) what you had in mind?

I'll then work on the following follow-up PRs if you agree:
- always rewrite literals even in non-matches (`+1` will be `{:integer, _, 1}` instead of `{:op, _, :+, {:integer, _, 1}}`)
- handle `0.0` in `Macro.escape/1` (necessary to fix the bug below)

```elixir
defmodule Foo do
  @x +0.0
  def foo(@x), do: 1
end

warning: pattern matching on 0.0 is equivalent to matching only on +0.0 from Erlang/OTP 27+. Instead you must match on +0.0 or -0.0
```

Reply:
Yes, that's what I had in mind. However, I am slightly confused with something: given we rewrite +0.0 AST into +0.0 literal, won't we always warn? Even when `def foo(+0.0)` is written? Since +0.0 (AST) will expand to -0.0 (literal)?

> handle 0.0 in Macro.escape/1 (necessary to fix the bug below)

Be careful with this. If +0.0 (literal) becomes +0.0 (AST), then expanding +0.0 (AST) cannot become +0.0 (literal), otherwise you may introduce an infinite loop. :)

Reply:
> handle 0.0 in Macro.escape/1 (necessary to fix the bug below)

Also, can we say from the code that the user intends to match only on +0.0/-0.0 or both? If you only have a value, you don't know the original context. Maybe the code was `@x some_expr_that_returns_float()`. So I don't think we should change Macro.escape yet!

Reply:
> Yes, that's what I had in mind. However, I am slightly confused with something: given we rewrite +0.0 AST into +0.0 literal, won't we always warn?

We don't since we go in the rewrite clause instead when we expand `{:., _, [Kernel, :+]}, _, [0.0]}`, we never expand `0.0`. There is no warning when running tests either (except the one I captured).

>  Even when `def foo(+0.0)` is written? 

Yup, it's fine as well 👍

<img width="978" alt="Screenshot 2023-09-22 at 23 36 35" src="https://github.com/elixir-lang/elixir/assets/11598866/1c4477b4-dcf8-4e4d-ac5a-3add1584e38b">


> Be careful with this. If +0.0 (literal) becomes +0.0 (AST), then expanding +0.0 (AST) cannot become +0.0 (literal), otherwise you may introduce an infinite loop. :)

Great point, I thought about a risk of infinite loops but maybe it's more dangerous than I thought, I'll be extra careful.

Reply:
> in your example, if some_expr_that_returns_float returns 0.0 which is +0.0, and we match on +0.0, this is not necessarily a bug but could be arguably the right thing to do?

The issue is that the code today matches both, so they need to decide if the intent is to match one or both. The whole issue is the ambiguity, there is no right answer unless we know if the value originally had a sign or not.